File: test_js_filter.py
import json
import re

with open('data.js', 'r', encoding='utf-8') as f:
    text = f.read()

match = re.search(r'window\.CONTRACT_DATA\s*=\s*(\[.*?\]);', text, re.DOTALL)
data = json.loads(match.group(1))

# replicate JS sort exactly
months_order = ['JAN','FEV','MAR','ABR','MAI','JUN','JUL','AGO','SET','OUT','NOV','DEZ']
all_keys = list(data[0].keys()) if len(data) > 0 else []

month_cols = []
for k in all_keys:
    if re.match(r'^[A-Z]{3}/\d{2}$', k):
        month_cols.append(k)

def sort_key(k):
    m, y = k.split('/')
    return (int(y), months_order.index(m))

month_cols.sort(key=sort_key)
print("monthColumns:", month_cols)

filter_month = 'FEV/26'
filter_idx = month_cols.index(filter_month)

month_start = '2026-02-01'
month_end = '2026-02-28'

emtel = None
for c in data:
    if c.get('SUBCONTRATADO') and 'EMTEL' in c['SUBCONTRATADO']:
        emtel = c
        break

if not emtel:
    print("EMTEL not found")
    exit()

def parse_date_str(d_str):
    if not d_str: return None
    d_str = str(d_str)
    m = re.match(r'^(\d{4})-(\d{2})-(\d{2})', d_str)
    if m:
        return f"{m.group(1)}-{m.group(2)}-{m.group(3)}"
    return d_str

d_inicio = parse_date_str(emtel.get('INICIO_CONTRATO'))
d_fim = parse_date_str(emtel.get('TERMINO_CONTRATO'))
print("dInicio:", d_inicio, "dFim:", d_fim, "monthStart:", month_start, "monthEnd:", month_end)

valid = True
if d_inicio and d_inicio > month_end: valid = False
if d_fim and d_fim < month_start: valid = False

print("Valid after date bounds:", valid)

def parse_num(val):
    if val is None: return 0
    if isinstance(val, (int, float)): return float(val)
    s = str(val).replace(' ', '').replace('R$', '')
    if ',' in s and '.' in s:
        s = s.replace('.', '').replace(',', '.')
    elif ',' in s:
        s = s.replace(',', '.')
    try:
        return float(s)
    except:
        return 0.0

baseAcumuladoNoMes = parse_num(emtel.get(filter_month))
prev_m_str = month_cols[filter_idx - 1] if filter_idx > 0 else None
prevM = parse_num(emtel.get(prev_m_str)) if prev_m_str else 0
tempMedido = baseAcumuladoNoMes - prevM

print(f"baseAcumuladoNoMes ({filter_month}):", baseAcumuladoNoMes, "prev_m_str:", prev_m_str, "prevM:", prevM, "tempMedido:", tempMedido)

didStartThisMonth = False
if d_inicio and (month_start <= d_inicio <= month_end):
    didStartThisMonth = True

print("didStartThisMonth:", didStartThisMonth)

if tempMedido <= 0 and not didStartThisMonth:
    valid = False

print("Final Valid:", valid)

File: Backup_16_04_2026/test_js_filter.js
const fs = require('fs');

const rawData = fs.readFileSync('data.js', 'utf8');
const jsonMatch = rawData.match(/window\.CONTRACT_DATA\s*=\s*(\[.*?\]);/s);
const data = JSON.parse(jsonMatch[1]);

const monthColumns = Object.keys(data[0] || {})
    .filter(k => /^[A-Z]{3}\/\d{2}$/.test(k))
    .sort((a, b) => {
        const months = ['JAN','FEV','MAR','ABR','MAI','JUN','JUL','AGO','SET','OUT','NOV','DEZ'];
        const [mA, yA] = a.split('/');
        const [mB, yB] = b.split('/');
        if (yA !== yB) return yA - yB;
        return months.indexOf(mA) - months.indexOf(mB);
    });

console.log("Month columns:", monthColumns.join(", "));

const filterMonth = 'FEV/26';
const filterMonthIndex = monthColumns.indexOf(filterMonth);
const monthMap = {'JAN':'01', 'FEV':'02', 'MAR':'03', 'ABR':'04', 'MAI':'05', 'JUN':'06', 'JUL':'07', 'AGO':'08', 'SET':'09', 'OUT':'10', 'NOV':'11', 'DEZ':'12'};
const parts = filterMonth.split('/');
const year = '20' + parts[1];
const monthNum = monthMap[parts[0]];
const daysInMonth = new Date(parseInt(year), parseInt(monthNum), 0).getDate();
const monthStart = `${year}-${monthNum}-01`;
const monthEnd = `${year}-${monthNum}-${daysInMonth}`;

const emtel = data.find(c => c['SUBCONTRATADO'] && c['SUBCONTRATADO'].includes('EMTEL'));

if (!emtel) {
    console.log("EMTEL not found!");
    process.exit(1);
}

const parseDateStr = (dateStr) => {
    if (!dateStr) return null;
    const dashMatch = String(dateStr).match(/^(\d{4})-(\d{2})-(\d{2})/);
    if (dashMatch) return `${dashMatch[1]}-${dashMatch[2]}-${dashMatch[3]}`;
    return String(dateStr);
};

const parseNumber = (val) => {
    if (typeof val === 'number') return isNaN(val) ? 0 : val;
    if (!val || val === '-' || val === ' ') return 0;
    let clean = val.toString().replace(/[\sR\$]/g, '');
    if (clean.includes(',') && clean.includes('.')) {
        clean = clean.replace(/\./g, '').replace(',', '.');
    } else if (clean.includes(',')) {
        clean = clean.replace(',', '.');
    }
    const num = parseFloat(clean);
    return isNaN(num) ? 0 : num;
};

// Simulate get contracts
let valid = true;
const dInicio = parseDateStr(emtel['INICIO_CONTRATO']);
const dFim = parseDateStr(emtel['TERMINO_CONTRATO']);
console.log("dInicio:", dInicio, "dFim:", dFim, "monthStart:", monthStart, "monthEnd:", monthEnd);

if (dInicio && dInicio > monthEnd) { console.log("Filtered by dInicio"); valid = false; }
if (dFim && dFim < monthStart) { console.log("Filtered by dFim"); valid = false; }

console.log("Valid after date bounds:", valid);

let baseAcumuladoNoMes = parseNumber(emtel[filterMonth]);
let prevMonthStr = filterMonthIndex > 0 ? monthColumns[filterMonthIndex - 1] : null;
let prevM = prevMonthStr ? parseNumber(emtel[prevMonthStr]) : 0;
let tempMedido = baseAcumuladoNoMes - prevM;

console.log("baseAcumuladoNoMes:", baseAcumuladoNoMes, "prevM:", prevM, "tempMedido:", tempMedido);

const didStartThisMonth = dInicio && (dInicio >= monthStart && dInicio <= monthEnd);
console.log("didStartThisMonth:", didStartThisMonth);

if (tempMedido <= 0 && !didStartThisMonth) {
    console.log("Filtered by zero production and NOT started this month");
    valid = false;
}

console.log("Final Valid:", valid);
